### ISSUE TITLE
Sample Chat Swift version - Array index out of bounds crash - Filtered Users method

### ISSUE THREAD

**Issue by pavithramouli** [QuickBlox/quickblox-ios-sdk#290]
Hi team, 

In the sample-chat-swift, retrieving the list of dev users result in a crash at 

`let filteredUsers = unsortedUsers[0..<kUsersLimit].filter { (user: QBUUser) -> Bool in`
in the function  `func filteredUsersByCurrentEnvironment() -> [QBUUser] `

I can get it to work only after commenting that section and returning the unsortedUsers list from the function. 
Logs show SDK as 2.6.5. Cloned your project.

**Comment by iSevenDays** [comment]
Thank you for reporting! We will check this and return to you soon

**Comment by iSevenDays** [comment]
I've checked that and I haven't found any problem. Could you please reinstall the application and run it again?

If you still have a crash, please attach it.

Thanks!

**Comment by pavithramouli** [comment]
[sample-chat-swift.zip](https://github.com/QuickBlox/quickblox-ios-sdk/files/126068/sample-chat-swift.zip)

Attached. 
I made changes in the App Delegate for application specific keys and modified test password in LoginTableViewController. Rest is as-is from the project sample you have.

**Comment by iSevenDays** [comment]
Seems that you have the same issue http://stackoverflow.com/questions/34898358/quickblox-ios-sdk-2-6-2-issue/34898852#34898852

**Comment by pavithramouli** [comment]
Not really. Actually I did have this error and added the dev tag to my users. That's fixed. 

Then, I have this crash. It shows a 
`fatal error: array index out of bounds ` exception and crashes at the line I have mentioned originally. My attached project doesn't crash for you? Weird!

**Comment by iSevenDays** [comment]
Our sample is based on user names

dev_user1
dev_user2

I was able to resolve the issue.

let filteredUsers = unsortedUsers[0..<kUsersLimit].filter { (user: QBUUser) -> Bool in
            return true
        }

Also please change

let kUsersLimit = 2


Thanks!

**Comment by pavithramouli** [comment]
Thanks. It doesn't crash with those two changes. 
So, I assume the actual array bounds in that function call should be the retrieved number of users and not the kUsersLimit right, in case of real world multi-person chat scenarios? 

And am not sure I understand the significance of the user names you have with the function implementation. Does it have any effect? Can you elaborate?

**Comment by iSevenDays** [comment]
> And am not sure I understand the significance of the user names you have with the function implementation. Does it have any effect? 

Yes, in our implementation we filter users with prefix dev_. But you can simply skip that.

> So, I assume the actual array bounds in that function call should be the retrieved number of users and not the kUsersLimit right, in case of real world multi-person chat scenarios?

Yes, you should not have kUsersLimit in your application and user array length information.

Thanks!

**Comment by pavithramouli** [comment]
Thanks!

**Comment by lenichols** [comment]
I seem to be having the same issue... how did you resolve it?  My application fails where yours did at this line:

`let filteredUsers = unsortedUsers[0..<kUsersLimit].filter { (user: QBUUser) -> Bool in
            return user.login?.lowercaseString.rangeOfString(containsString) != nil
        }`

What did you do to resolve this @pavithramouli @iSevenDays

**Comment by pavithramouli** [comment]
Hi @lenichols , what I understood from the previous comments here was that the `kUsersLimit` constant was set for their demo data set which had 10 users. I was testing with just 2 in my environment and so I had this crash. I assume, your case is similar too. 

This filtering logic is also custom for their users, so simple solution for you is to comment out that parts dealing with `filteredUsers` section and just do `return unsortedUsers`

If you want the demo to work as is, you need to set the `kUsersLimit` to the actual number of users you are testing with. Also, don't forget to add tag to your users as '_dev_'.

**Comment by lenichols** [comment]
@pavithramouli can you attach your ServicesManager.swift file so I can see? When i comment filtered users out the app won't run. Also, how do you set the app to production mode? I am trying to modify it but having to tag all users for dev is kind of over kill.

**Comment by pavithramouli** [comment]
Here is the function that works for me in the sample project. 
`

      func filteredUsersByCurrentEnvironment() -> [QBUUser] { 
        let unsortedUsers = self.usersService.usersMemoryStorage.unsortedUsers() as! [QBUUser]
        return unsortedUsers
    }


`


Like you see, I am returning just the `unsortedUsers` array and have commented everything else from the function. Works for me with two sample users of my own. 

I can think of a few cases why you can't run your project. See if any of these help. 

- there have been a few commits to the project, probably since the time you cloned, try and clean your project folder of the `Pods directory, .lock file and the .xcworkspace`. Copy the new podfile contents from the latest commit and then run a fresh `pod install` 
- you need to specify your user password in `LoginTableViewController` for constant `kTestUsersDefaultPassword`. Originally it is with their sample users. 
- you need to have your own app-specific keys in `AppDelegate`

As far as `dev` tag is concerned, I think the QuickBlox developers can answer better. I can only assume that this distinction helps differentiate different stages of development - like dev, qa, production etc. I agree that the distinction could have been handled differently or at least the current process documented better.

**Comment by lenichols** [comment]
@pavithramouli thank you! thank you! thank you!... so much for this... your function works just fine and my app loads without issue now. Now if only someone like @iSevenDays or @Raikerian can tell me how to get the app out of dev and into prod i should be good to go.

**Comment by pavithramouli** [comment]
Good to know! :) 

And what exactly is your problem for having your app run in production mode? Am not there yet but I don't see any settings that distinguish the app tokens to dev or prod. Is QuickBlox blocking you in any way? Or do you want clarity on push notifications? You may get better answers if you open a new issue for this problem (if there is any).
